sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -1702,20 +1696,33 @@ pi_result _pi_ze_event_list_t::createAndRetainPiZeEventList(
         }
 
         auto Queue = EventList[I]->Queue;
-
-        if (Queue && Queue != CurQueue) {
-          // If the event that is going to be waited on is in a
-          // different queue, then any open command list in
-          // that queue must be closed and executed because
-          // the event being waited on could be for a command
-          // in the queue's batch.
-
-          // Lock automatically releases when this goes out of scope.
-          std::scoped_lock lock(Queue->Mutex);
-
-          if (auto Res = Queue->executeAllOpenCommandLists())
-            return Res;
-        } else if (!Queue) {
+        if (Queue) {
+          // The caller of createAndRetainPiZeEventList must already hold
+          // a lock of the CurQueue. Additionally lock the Queue if it
+          // is different from CurQueue.
+          auto Lock = ((Queue == CurQueue) ? std::unique_lock<pi_shared_mutex>()
+                                           : std::unique_lock(Queue->Mutex));
+
+          // If the event that is going to be waited is in an open batch
+          // different from where this next command is going to be added,
+          // then we have to force execute of that open command-list
+          // to avoid deadlocks.
+          //
+          const auto &OpenCommandList =
+              Queue->eventOpenCommandList(EventList[I]);
+          if (OpenCommandList != Queue->CommandListMap.end()) {
+
+            if (Queue == CurQueue &&
+                OpenCommandList->second.isCopy(Queue) == UseCopyEngine) {
+              // Don't force execute the batch yet since the new command
+              // is going to the same open batch as the dependent event.
+            } else {
+              if (auto Res = Queue->executeOpenCommandList(
+                      OpenCommandList->second.isCopy(Queue)))
+                return Res;
+            }
```

Comment:
From this code it looks like we always force execute open command list if it is compute command list. Could you please explain why  behavior needs to be different for copy and compute command lists? I think comment https://github.com/intel/llvm/pull/6199/files#diff-15dd1eb076d2164bd9e87d9057f05f652a716498e8cdf5975e564c65309a0985R1706-R1709 should mention about this difference.

Reply:
It is not different for copy vs. compute. This code force-submits copy if the dependent command is compute, and vice versa. Perhaps, you were confused by this check `OpenCommandList->second.isCopy(Queue) == UseCopyEngine` looking like it is about copy-only, but I think it is not :)

Reply:
Oh, I see, yes, I was confused by this check.